README.md
# intonation

### [R](https://www.r-project.org/) package for the linguistic analysis of fundamental frequency (F0) in speech

Version 1.1 (July 12, 2016)

*[Aaron Albin](http://www.aaronalbin.com/)*

## What this package does

This package consists of the functions used in [Albin (2015)](http://hdl.handle.net/2022/20345) for various tasks relating to speech [fundamental frequency](https://en.wikipedia.org/wiki/Fundamental_frequency) (F0), including:

1. manipulating Pitch objects from the phonetics software [Praat](http://www.fon.hum.uva.nl/praat/),
2. visualizing F0 data in a way that retains rich information from the acoustic signal, and
3. creating and plotting a schematic quantitative representation ('stylization') of an F0 track.

This package is referenced on page 52 of:

> Albin, A. (2015). *Typologizing native language influence on intonation in a second language: Three transfer phenomena in Japanese EFL learners*. (Doctoral dissertation). Indiana University, Bloomington.[http://dx.doi.org/10.5967/K8JW8BSC](http://hdl.handle.net/2022/20345)

The [data used in the dissertation](http://hdl.handle.net/2022/20346) are also available for download.


While at present the contents of this package are limited to the functions from the dissertation, it is anticipated that this package will grow over time to incorporate a variety of other intonation-related functions. As such, other prosody researchers are more than welcome to contribute their own R code to this package. (See the [author's homepage](http://www.aaronalbin.com/) for contact information.)

## Installation

This package has two other package dependencies you must install first.

1. First, download the package [audio](https://cran.r-project.org/web/packages/audio/index.html) from CRAN, e.g. with:
    install.packages("audio").

2. Second, install PraatR by going to the [PraatR homepage](http://www.aaronalbin.com/praatr/) and following the steps under "Installation".

Now you can install this package by following these steps:

1. Click on the "Download ZIP" button on the right side of this page and save the .zip file somewhere on your computer.

2. Unzip the .zip file to somewhere convenient (e.g. your desktop).

 * After this step, the .zip file is no longer needed and can be deleted.

3. Open R.

 * If you are running Windows, you must open R as an administrator. To do so, go to the icon for R on your start menu ('start screen' in Windows 8) or desktop, right-click it, and select "Run as administrator". Click "Yes" to the User Account Control window that pops up.

4. Find the path to the the unzipped folder from step 2. If you unzipped to your desktop, the path will be something like the following:

 - **Windows**: "C:/Users/MyUsername/Desktop/intonation-master/"

 - **Mac**: "/Users/MyUsername/Desktop/intonation-master/"

5. Run the following line of code, adjusting the path for the *pkgs* argument as needed so it points to path you determined in Step 4.

        install.packages(pkgs="(path from Step 4)", repos=NULL, type="source")

If everything works correctly, you should see something like the following in the R console:

    * installing *source* package 'intonation' ...
    ** R
    ** inst
    ** preparing package for lazy loading
    ** help
    *** installing help indices
    ** building package indices
    ** testing if installed package can be loaded
    *** arch - i386
    *** arch - x64
    * DONE (intonation)

Once installed, type *?intonation* at the R console to pull up the help file for the package, which acts as a table of contents directing you to help files for other specific functions. You can also type *example("intonation")* to see some examples of what the package can do.

## Contents

In addition to some sample data ([HelloWorld](https://github.com/usagi5886/intonation/blob/master/inst)), the package includes the following functions

 - [Spectrogram()](https://github.com/usagi5886/intonation/blob/master/R/Spectrogram.r): Create a spectrogram of an audio file
 - [F0RangeFinder()](https://github.com/usagi5886/intonation/blob/master/R/F0RangeFinder.r): Cycle through each soundfile in a folder and determine F0 range for each
 - [ToPitch()](https://github.com/usagi5886/intonation/blob/master/R/ToPitch.r): Use Praat to generate a Pitch object
 - [ReadPitch()](https://github.com/usagi5886/intonation/blob/master/R/ReadPitch.r): Read a Pitch object into R
 - [RichVisualization()](https://github.com/usagi5886/intonation/blob/master/R/RichVisualization.r): Visualize the F0 data in a Pitch object
 - [Stylize()](https://github.com/usagi5886/intonation/blob/master/R/Stylize.r): Create a 'stylization' (i.e., a schematic representation) of F0 an track
 - [PlotStylization()](https://github.com/usagi5886/intonation/blob/master/R/PlotStylization.r): Draw a stylization on an open plot of an F0 track

Each of these has a help file included in the package. Thus, for example, you can type *?Stylize* to find out more information about how to use the Stylize() function.

## Usage

R documentation 'help files' are available for every function, but the following gives as a rough idea about what the syntax for these functions looks like:

    # Draw a spectrogram of the audio data in a soundfile
    Spectrogram(WavePath)

    # Quickly determine F0 range for every soundfile in one folder (WaveFolder)
    # Save the resulting pitch-related files in another folder (PitchFolder)
	F0Ranges = F0RangeFinder( WaveFolder="C:/Wave/", PitchFolder="C:/Pitch/")

    # Use an F0 range thus determined to create a Pitch object from this soundfile
    ToPitch( Input=WavePath, Output=PitchPath, Range=c(69,135) )

    # Read this Pitch object into R
    ReadPitch( PitchPath )

    # Plot the F0 data contained therein as a 'rich visualization'
    RichVisualization( PitchPath=PitchPath, WavePath=WavePath, Labels = c("hello","world"), Divisions_ms = c(132,648,1257) )

    # Create a stylization of the F0 track
    Stylization = Stylize( PitchPath, VertexIndices=c(211,489,1123) )

    # Superimpose this stylization on the rich visualization
    PlotStylization(Stylization)

## License

This package is released under the GNU General Public License:

> This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

> This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.  See the GNU General Public License for more details.

> To receive a copy of the GNU General Public License, see: [http://www.gnu.org/licenses/](http://www.gnu.org/licenses/)


NEWS
NEWS in package 'intonation'

1.1	07-12-2016
    o	Added F0RangeFinder() function

1.0	08-30-2015
    o	Initial public release


R/ReadPitch.r
##################################
# ReadPitch()                    #
# ------------------------------ #
# Copyright (C) 2015 Aaron Albin #
# http://www.aaronalbin.com/     #
##################################

# Takes a Pitch object textfile, reads it into R in a quick and efficient way, and returns the requested information.

# This function is referenced on page 55 of:
# Albin, A. (2015). Typologizing native language influence on intonation in
#     a second language: Three transfer phenomena in Japanese EFL learners.
#     (Doctoral dissertation). Indiana University, Bloomington.
#     http://dx.doi.org/10.5967/K8JW8BSC

ReadPitch = function( # Begin argument list

# [1] File path
File, # Path to a Pitch object stored on one's computer
# This must be stored in Praat's "text" format (and not "short (text)" or "binary" format).
# This does *not* have to be a full path, and it *can* have spaces in it.

# [2] 'Return'
Return="Frames" # Specifies the kind of information about the Pitch object to be brought into R.
# At present, three values are accepted:
# - 'Header': All the info from the header, in the same order (but with more transparent names)
# - 'Frames': Information about each frame (and each frame's highest-ranking candidate)
# - 'Candidates': Information about every candidate in each frame
# The default is 'Frames' since it represents the most common use case.

){ # Begin body of function

# First, do some basic input validation
# Check whether a file indeed exists at the specified file path
if( !file.exists(File) ){stop("The specified file does not exist.")}
# Note that it is assumed (but not explicitly verified) that this file is indeed a Pitch object.

# Make sure the value supplied for 'Return' is a permissible choice
if( !( Return %in% c("Header","Frames","Candidates") ) ){
	stop("The argument 'Return' must be set to 'Header', 'Frames', or 'Candidates'.")
} # End 'if' statement

# Now proceed with the function proper...

# Read in the file
if(Return=="Header"){ # If 'Return' is 'Header'...
	nLines=10 # ...read in just the first 10 lines (to save time).
}else{ # Otherwise, read in the whole thing.
	nLines=-1 # -1 is the default for readLines(), telling R to read in the whole file
} # End 'if/else'
FullText = readLines(File,n=nLines)

# Now extract all of the information from the header (i.e. the first 10 lines of the Pitch object).
# The information in the header looks like the following:

# [01] File type = "ooTextFile"
# [02] Object class = "Pitch 1"
# [03] 
# [04] xmin = 0 
# [05] xmax = 3.3430625 
# [06] nx = 331 
# [07] dx = 0.01 
# [08] x1 = 0.02153125000000011 
# [09] ceiling = 600 
# [10] maxnCandidates = 15 

FileStartTime       = as.numeric(strsplit(FullText[ 4],split=" ")[[1]][3])
FileEndTime         = as.numeric(strsplit(FullText[ 5],split=" ")[[1]][3])
nFrames             = as.integer(strsplit(FullText[ 6],split=" ")[[1]][3])
TimeStep            = as.numeric(strsplit(FullText[ 7],split=" ")[[1]][3])
FirstTime           = as.numeric(strsplit(FullText[ 8],split=" ")[[1]][3])
Ceiling             = as.integer(strsplit(FullText[ 9],split=" ")[[1]][3])
MaxNumberCandidates = as.integer(strsplit(FullText[10],split=" ")[[1]][3])

if(Return=="Header"){ # Begin outer if/else...

# Return the information from the header in the form of a list
# The list format ensures everything prints 'pretty' (e.g. with integers treated as integers).
# Moreover, the entries in the list can still be easily accessed with '$'.
return( list( FileStartTime=FileStartTime,
                            FileEndTime=FileEndTime,
                            nFrames=nFrames,
                            TimeStep=TimeStep,
                            FirstTime=FirstTime,
                            Ceiling=Ceiling,
                            MaxNumberCandidates=MaxNumberCandidates ) )
# Note that the components of this list are in the same order as in the actual header itself (just with different names)

# Note that, unfortunately, the floor is *not* stored inside the header of a Pitch object.
# It can be merely guessed at from lowest observed F0 candidate as follows:
#ApproximateFloor = round( min(F0Values,na.rm=TRUE) ) # Not exact!
# (Here, 'round()' is used instead of 'floor' because sometimes sampled F0 values can go under an actual specified floor a tiny bit.)

}else{ # i.e. if 'Return' is *not* 'Header'... (Outer if/else)

# Use the information from the header to reconstruct the time series
Time=FirstTime+TimeStep*0:(nFrames-1)

# Extract out the frequency of each candidate
FrequencyLines = grep(FullText,pattern="                frequency = ",fixed=TRUE,value=TRUE)
Frequency = as.numeric( substring(FrequencyLines,first=29,last=nchar(FrequencyLines)-1) )
Frequency[Frequency == 0] <- NA # Replace dummy '0' with (the more semantically transparent) NA

# Extract out the strength of each candidate
StrengthLines = grep(FullText,pattern="                strength = ",fixed=TRUE,value=TRUE)
Strength = as.numeric( substring(StrengthLines,first=28,last=nchar(StrengthLines)-1) )
Strength[Strength == 0] <- NA # Again, replace dummy '0' with NA.

# Extract out the intensity of each frame
IntensityLines = grep(FullText,pattern="        intensity = ",fixed=TRUE,value=TRUE)
Intensity = as.numeric( substring(IntensityLines,first=21,last=nchar(IntensityLines)-1) )

# Extract out the number of candidates in each frame
nCandidatesLines = grep(FullText,pattern="        nCandidates = ",fixed=TRUE,value=TRUE)
nCandidates = as.integer( substring(nCandidatesLines,first=23,last=nchar(nCandidatesLines)-1) )

# Use the 'nCandidates' information to construct a 'nthCandidate' vector
nthCandidate = unlist( lapply( 1:nFrames, FUN=function(x){1:nCandidates[x]}))

# Use all of the information assembled above to filter down the candidate-level Frequency and Strength vectors into two smaller vectors containing just the information for the first (i.e. highest-ranked) candidates
Frequency1 = Frequency[nthCandidate==1]
Strength1 = Strength[nthCandidate==1]
# Note: The '1' suffix in the variable names indicates 'ranked 1st'.

if(Return=="Frames"){ # Begin inner if/else...
# Return the information about each frame in the form of a dataframe
return( data.frame( # nthFrame=1:nFrames, # Excluded since largely redundant (and easy to reconstruct)
					Time=Time,
                    Frequency=Frequency1, # Note name change
                    Intensity=Intensity,
                    Strength=Strength1, # Note name change
                    nCandidates=nCandidates ) )

}else{ # i.e. if 'Return' is *not* 'Frames'... (Inner if/else)
# By this point, by virtue of the the input validation at the beginning of the function, it is safely guaranteed that 'Return' is 'Candidates'.

# Create an 'nthFrame' vector, indexing which frame the measurements are associated with
nthFrame = unlist( lapply( 1:nFrames, FUN=function(x){rep(x, times=nCandidates[x])}) )

# Return the information about the candidates in each frame in the form of a dataframe
return( data.frame( nthFrame=nthFrame,
                    nthCandidate=nthCandidate,
                    Frequency=Frequency,
                    Strength=Strength ) )

} # End inner if/else ('Return' is 'Frames'?)

} # End outer if/else ('Return' is 'Header'?)

###################
# Arguments:      #
###################
# File            #
# Return="Frames" #
###################

} # End definition of function 'ReadPitch()'

########################################################################
# This program is free software: you can redistribute it and/or modify #
# it under the terms of the GNU General Public License as published by #
# the Free Software Foundation, either version 3 of the License, or    #
# (at your option) any later version.                                  #
#                                                                      #
# This program is distributed in the hope that it will be useful,      #
# but without any warranty; without even the implied warranty of       #
# merchantability or fitness for a particular purpose.  See the        #
# GNU General Public License for more details.                         #
#                                                                      #
# To receive a copy of the GNU General Public License, see:            #
# http://www.gnu.org/licenses/                                         #
########################################################################


R/PlotStylization.r
##################################
# PlotStylization()              #
# ------------------------------ #
# Copyright (C) 2015 Aaron Albin #
# http://www.aaronalbin.com/     #
##################################

# This function takes a Stylization dataframe as input and draws it the information therein (e.g. the vertex locations and transition shapes) to draw the stylization in R's currently active window.
# It is assumed that this window contains an F0 track (x axis = time, y axis = F0) representing the same soundfile that the stylization is based on.
# This F0 track plot will generally be a rich visualization, created with the RichVisualization() function in this package, but PlotStylization() will work even if this is not the case.
# The stylization will be superimposed on top of this F0 track in the form of points and lines.

# This function is referenced on page 67 of:
# Albin, A. (2015). Typologizing native language influence on intonation in
#     a second language: Three transfer phenomena in Japanese EFL learners.
#     (Doctoral dissertation). Indiana University, Bloomington.
#     http://dx.doi.org/10.5967/K8JW8BSC

PlotStylization = function( # Begin argument list

Stylization # A stylization dataframe created with the Stylize() function

){ # End argument list; begin main function body

# Check to see whether a graphics device is open.
# If not, stop computation.
if( is.null( dev.list() ) ){ stop("No graphics devices are open. Create the F0 track plot first.") }

# Count how many transitions are represented in the Stylization
nTransitions = nrow(Stylization)-1
# Recall how any given vertex is included in the set of F0 points for both the movement to the left and the movement to the right.

# Thus paired into transitions, loop through the rows
for(EachTransition in 1:nTransitions){ # EachTransition=1

	#______#
	# TIME #

	# Extract the timestamps at the beginning and end of this transition
	CurrentStartingTime = Stylization[EachTransition,"Time_ms"]
	CurrentEndingTime = Stylization[EachTransition+1,"Time_ms"]

	# Use this information to make a series of 100 points across the transition
	# (The choice of the number 100 is of course arbitrary, but it suffices for making curves of sufficient resolution for plotting purposes.)
	TimeSeries = seq(from=CurrentStartingTime,to=CurrentEndingTime,length.out=100)

	# Also create a normalized version of this spanning from 0 to 1
	NormalizedTime = seq(from=0,to=1,length.out=100)

	#____#
	# F0 #

	# Determine the F0 values at the beginning and end of this transition
	CurrentStartingF0 = Stylization[EachTransition,"F0"]
	CurrentEndingF0 = Stylization[EachTransition+1,"F0"]

	# Calculate the magnitude of the resulting F0 movement
	RangeF0Movement = abs(CurrentStartingF0-CurrentEndingF0)

	# Set up a logical variable indicating whether this F0 movement is rising (TRUE) or falling (FALSE)
	RisingMovement = CurrentEndingF0 > CurrentStartingF0

	#_____________________#
	# THRESHOLD/GRADIENCE #

	# Extract out the threshold and gradience parameters associated with this transition
	CurrentThreshold = Stylization[EachTransition+1,"Threshold"]
	CurrentGradience = Stylization[EachTransition+1,"Gradience"]

	# Using the equations in (4) on p.79 of Albin (2015), convert these into the shape1 ('A') and shape2 ('B') parameters of the beta distribution
	shape1=(2*   CurrentThreshold )*CurrentGradience+1
	shape2=(2*(1-CurrentThreshold))*CurrentGradience+1
	# Solved in the opposite direction:
	# Gradience=(A/2)+(B/2)-1 # This can be simplified to Gradience=( (A+B)/2 ) - 1
	# Threshold=(A-1)/(A+B-2) # In more transparent form: Threshold=(A-1)/( (A-1) + (B-1) )

	#____________________#
	# ASSEMBLING CONTOUR #

	# Run the two shape parameters through pbeta() - the cumulative beta distribution function - to generate a curve shape
	BetaVector = pbeta(NormalizedTime, shape1=shape1, shape2=shape2)

	if(RisingMovement){ # If the F0 movement is a rise... 
		FlippedBeta = BetaVector # ...leave the curve shape as is
	}else{ # If, instead, it is a fall...
		FlippedBeta = ( -1 * BetaVector ) # ...flip it vertically (to make the curve shape a fall)
	}

	# Put the abstract curve shape into actual F0 coordinates (by adding the starting F0 value and stretching vertically by the size of the F0 movement)
	y = CurrentStartingF0 + (RangeF0Movement*FlippedBeta) 

	# Extract out from the Stylization dataframe what percent of the frames in this transition are voiced
	CurrentPercentVoiced = Stylization[EachTransition+1,"PercentVoiced"]

	# Bring everything up to this point together in a call to lines() to draw the transition lines
	lines( x=TimeSeries,
		   y=y,
		   col=grey(1-CurrentPercentVoiced), # In grey(), values closer to 0 are blacker, and values closer to 1 are whiter
		   lwd=CurrentPercentVoiced*2
	) # End call to lines()

} # End 'each transition'

# Finally, add the points for the vertices themselves
# This is performed at this point in the code (and not inside the for-loop above) so that the dots for the vertices appear *on top of* the transition lines.
points(x=Stylization$Time_ms, y=Stylization$F0, pch=21, bg="white",cex=1.5)

###############
#  ARGUMENTS  #
###############
# Stylization #
###############

} # End definition of function 'PlotStylization()'

########################################################################
# This program is free software: you can redistribute it and/or modify #
# it under the terms of the GNU General Public License as published by #
# the Free Software Foundation, either version 3 of the License, or    #
# (at your option) any later version.                                  #
#                                                                      #
# This program is distributed in the hope that it will be useful,      #
# but without any warranty; without even the implied warranty of       #
# merchantability or fitness for a particular purpose.  See the        #
# GNU General Public License for more details.                         #
#                                                                      #
# To receive a copy of the GNU General Public License, see:            #
# http://www.gnu.org/licenses/                                         #
########################################################################


R/ToPitch.r
##################################
# ToPitch()                      #
# ------------------------------ #
# Copyright (C) 2015 Aaron Albin #
# http://www.aaronalbin.com/     #
##################################

# Takes a soundfile and produces a Pitch object using the specified input parameters

# This function is referenced on page 55 of:
# Albin, A. (2015). Typologizing native language influence on intonation in
#     a second language: Three transfer phenomena in Japanese EFL learners.
#     (Doctoral dissertation). Indiana University, Bloomington.
#     http://dx.doi.org/10.5967/K8JW8BSC

# _________________________________________________________________________
# _________________________________________________________________________
# _________________________________________________________________________

# Praat has five different commands that extract F0:
# [1] To Pitch...
# [2] To Pitch (ac)...
# [3] To Pitch (cc)...
# [4] To Pitch (SPINET)...
# [5] To Pitch (shs)...

# Of these, [1] is just a high-level convenience wrapper/interface to [2].

# The three arguments to "Sound: To Pitch", along with their defaults, are as follows
# Time step (s):        0.0 (= auto)
# Pitch floor (Hz):    75.0
# Pitch ceiling (Hz): 600.0

# If you use "To Pitch (ac)...", you unlock all of these advanced settings:
# (See also Table 3.2 on p56 of Albin (2015).)

# Finding the candidates
#   Time step (s):              0.0 (= auto)
#   Pitch floor (Hz):          75.0
#   Max. number of candidates: 15
#   [+/-] Very accurate        (In PraatR, "no" or "yes")
# Finding a path:
#   Silence threshold:         0.03
#   Voicing threshold:         0.45
#   Octave cost:               0.01
#   Octave-jump cost:          0.35
#   Voiced / unvoiced cost:    0.14
#   Pitch ceiling (Hz):      600.0  # NOTE: The default is 500 in the editor window.

# In order to maximum flexibility by unlocking all of these advance settings, this function uses "To Pitch (ac)...".
# The decrement in speed relative to "To Pitch..." is probably negligible.

ToPitch = function( # Begin argument list

# [1-2] Input and output
Input, # Path to a WAV file on one's computer
Output, # Path to the Pitch object textfile that will be created as output
# Neither of these arguments have defaults.
# These must both be full file paths (i.e., you can't rely on R's working directory functionality)
# The file paths also must contain no spaces (due to a limitation in the current implementation of PraatR).
# Note that, unlike the 'PraatR' package, both of these argument names are capitalized ('[I]nput' and '[O]utput')

# [3] Overwrite
Overwrite=FALSE, # If the 'Output' file already exists on your computer, determines whether it will be overwritten.
# To be on the safe side, the default is left at FALSE (following PraatR), whereby an error message is issued.
# Here again, this argument name is capitalized ('[O]verwrite'), unlike in PraatR.
# This is passed directly to the 'praat()' function.

# [4] FileType
FileType = "text", # What kind of format the Pitch object data should be stored in
# The options are 'text', 'short text' (='short'), or 'binary'.
# 'binary' saves the most disk space, whereas 'text' (the default) is more convenient for processing in R.

# [5] Range
Range, # The F0 range (in Hertz) for the F0 analysis, specified as a vector
# Either c(ceiling, floor) or c(floor, ceiling) will work, but the latter is perhaps more conventional.
# In Praat, the default floor is 75 and the default ceiling is either 500 (objects window) or 600 (editor window).
# In this function, the 'Range' argument does *not* have a default value, though, since it is good practice to specify the range on a file-by-file basis.
# (Even if the same value is used for many files, specifying it makes one's assumptions explicit.)

# [6] Time step
TimeStep=0.001, # Size of the time step (or 'frame') for the analysis, in milliseconds
# The default of 0.001 means an F0 value will be extracted at every millisecond.
# Note that this is *not* the Praat default, which is '0' (auto).

# [7] Maximum number of candidates
MaxNumberOfCandidates=15, # The maximum number of candidate F0 values in each frame
# The default value here (15) follows Praat's default.
# Setting this to 2 will save disk space and will not affect many (if not most) analyses.

# [8] Advanced parameters
# All of these are left at their defaults and therefore normally do not need to be specified.
# When they do need to be specified, they can just be included in the call to ToPitch() as needed.
VeryAccurate=FALSE, # Note: 'FALSE' (logical) is interchangeable with "no" (character) in PraatR.
SilenceThreshold=0.03,
VoicingThreshold=0.45,
OctaveCost=0.01,
OctaveJumpCost=0.35,
VoicedUnvoicedCost=0.14

){ # End argument list, and begin body of function definition

# Require the PraatR package
require("PraatR")

# Check to make sure none of the required arguments are missing:
MissingCheck = missing(Input) | missing(Output) | missing(Range)
if( MissingCheck ){
	MissingList = paste( c("Input", "Output", "Range")[ c( missing(Input), missing(Output), missing(Range) ) ], collapse=", " )
	stop(paste("You need to specify the following arguments: ",MissingList))
} # End 'if any necessary arguments are missing'

# Do some basic input validation to ensure Range is a numeric (or integer) vector of length 2
if(!( mode(Range) %in% c("numeric","integer") )){ stop("'Range' must be a numeric or integer vector.") }
if(length(Range)!=2){ stop("'Range' must be of length 2, e.g. c(floor,ceiling)") }
FinalRange = sort(Range) # to ensure the F0 range is in the standard format c(floor, ceiling)

# Assemble all of the arguments into a list
Arguments=list(TimeStep=TimeStep,
               PitchFloor=FinalRange[1],
               MaxNumberOfCandidates=MaxNumberOfCandidates,
               VeryAccurate=VeryAccurate,
               SilenceThreshold=SilenceThreshold,
               VoicingThreshold=VoicingThreshold,
               OctaveCost=OctaveCost,
               OctaveJumpCost=OctaveJumpCost,
               VoicedUnvoicedCost=VoicedUnvoicedCost,
               PitchCeiling=FinalRange[2])
# Note that the ordering in this list follows Praat's order of arguments.
# This is different from the default order of arguments to this function (ToPitch()) as a whole.

# Perform the actual PraatR command
# This requires PraatR to be loaded, but it is a package dependency for the 'intonation' package as a whole
PraatR:::praat( command="To Pitch (ac)...",
                arguments=Arguments,
                input=Input,
                output=Output,
                overwrite=Overwrite,
                filetype=FileType
) # End call to 'praat()'

############################
#    A R G U M E N T S     #
############################
# Input                    #
# Output                   #
# Overwrite=FALSE          #
# FileType="text"          #
# Range                    #
# TimeStep=0.001           #
# MaxNumberOfCandidates=15 #
# VeryAccurate=FALSE       #
# SilenceThreshold=0.03    #
# VoicingThreshold=0.45    #
# OctaveCost=0.01          #
# OctaveJumpCost=0.35      #
# VoicedUnvoicedCost=0.14  #
############################

} # End definition of function 'ToPitch()'

########################################################################
# This program is free software: you can redistribute it and/or modify #
# it under the terms of the GNU General Public License as published by #
# the Free Software Foundation, either version 3 of the License, or    #
# (at your option) any later version.                                  #
#                                                                      #
# This program is distributed in the hope that it will be useful,      #
# but without any warranty; without even the implied warranty of       #
# merchantability or fitness for a particular purpose.  See the        #
# GNU General Public License for more details.                         #
#                                                                      #
# To receive a copy of the GNU General Public License, see:            #
# http://www.gnu.org/licenses/                                         #
########################################################################
